Docs: Grammar action, satisfaction! First 20 pages. (#1817)
dlt-hub/dlt: docs/website/docs/general-usage/incremental-loading.md

</div>

> The "write disposition" you choose depends on the dataset and how you can extract it.

Reviewer: This is a strange change as I think both terms are valid

Author: We use dataset throughout the docs and in the code, so I would say it is good, no?

2. `scd2`
3. `upsert`

> ### `Delete-insert` strategy

Reviewer: ```suggestion
### `delete-insert` strategy
```
Should keep backticked strings unchanged

- `mssql`
- `postgres`
- `snowflake`

> - 🧪 `filesystem` with `delta` table format (see limitations [here](../dlt-ecosystem/destinations/filesystem.md#known-limitations))

Reviewer: ```suggestion
- `filesystem` with `delta` table format (see limitations [here](../dlt-ecosystem/destinations/filesystem.md#known-limitations))
```
Lets remove the emoji as its meaning could be unclear in this context

In essence, the `dlt.sources.incremental` instance above:
* **updated_at.initial_value** which is always equal to "1970-01-01T00:00:00Z" passed in the constructor
* **updated_at.start_value** a maximum `updated_at` value from the previous run or the **initial_value** on the first run
* **updated_at.last_value** a "real-time" `updated_at` value updated with each yielded item or page. Before the first yield, it equals **start_value**
* **updated_at.end_value** (here not used) [marking the end of the backfill range](#using-end_value-for-backfill)

When paginating, you probably need the **start_value** which does not change during the execution of the resource, however, most paginators will return a **next page** link which you should use.

> Behind the scenes, `dlt` will deduplicate the results, i.e., in case the last issue is returned again (`updated_at` filter is inclusive) and skip already loaded ones.

Reviewer: ```suggestion
Behind the scenes, dlt will deduplicate the results, i.e., in case the last issue is returned again (`updated_at` filter is inclusive) and skip already loaded ones.
```

        yield page
```

We just yield all the events and `dlt` does the filtering (using the `id` column declared as `primary_key`).

GitHub returns events ordered from newest to oldest. So we declare the `rows_order` as **descending** to [stop requesting more pages once the incremental value is out of range](#declare-row-order-to-not-request-unnecessary-data). We stop requesting more data from the API after finding the first event with `created_at` earlier than `initial_value`.

:::note

> `dlt.sources.incremental` is implemented as a [filter function](resource.md#filter-transform-and-pivot-data) that is executed **after** all other transforms you add with `add_map` / `add_filter`. This means that you can manipulate the data item before the incremental filter sees it. For example:

Reviewer: ```suggestion
`dlt.sources.incremental` is implemented as a [filter function](resource.md#filter-transform-and-pivot-data) that is executed **after** all other transforms you add with `add_map` or  `add_filter`. This means that you can manipulate the data item before the incremental filter sees it. For example:
```

...
```

> Note that `dlt`'s incremental filtering considers the ranges half-closed. `initial_value` is inclusive, `end_value` is exclusive, so chaining ranges like above works without overlaps.

Reviewer: ```suggestion
Note that dlt's incremental filtering considers the ranges half-closed. `initial_value` is inclusive, `end_value` is exclusive, so chaining ranges like above works without overlaps.
```

### Declare row order to not request unnecessary data

With the `row_order` argument set, `dlt` will stop retrieving data from the data source (e.g., GitHub API) if it detects that the values of the cursor field are out of the range of **start** and **end** values.

In particular:

> * `dlt` stops processing when the resource yields any item with a cursor value _equal to or greater than_ the `end_value` and `row_order` is set to **asc**. (`end_value` is not included)
> * `dlt` stops processing when the resource yields any item with a cursor value _lower_ than the `last_value` and `row_order` is set to **desc**. (`last_value` is included)

Reviewer: ```suggestion
* dlt stops processing when the resource yields any item with a cursor value _equal to or greater than_ the `end_value` and `row_order` is set to **asc**. (`end_value` is not included)
* dlt stops processing when the resource yields any item with a cursor value _lower_ than the `last_value` and `row_order` is set to **desc**. (`last_value` is included)
```

    ):
        yield page
```

We opt-in to the Airflow scheduler by setting `allow_external_schedulers` to `True`:

> 1. When running on Airflow, the start and end values are controlled by Airflow and the `dlt` [state](state.md) is not used.

Reviewer: ```suggestion
1. When running on Airflow, the start and end values are controlled by Airflow and the dlt [state](state.md) is not used.
```

> 2. In all other environments, the `incremental` behaves as usual, maintaining the `dlt` state.

Reviewer: ```suggestion
2. In all other environments, the `incremental` behaves as usual, maintaining the dlt state.
```
